commands/command_merge_driver.go

```diff
@@ -0,0 +1,140 @@
+	defer mergeCleanup(replacements)
+
+	var exitStatus int
+	formattedMergeProgram := subprocess.FormatPercentSequences(mergeDriverProgram, replacements)
+	cmd, err := subprocess.ExecCommand("sh", "-c", formattedMergeProgram)
+	if err != nil {
+		return -1, errors.New(tr.Tr.Get("failed to run merge program %q: %s", formattedMergeProgram, err))
+	}
+	err = cmd.Run()
+	// If it runs but exits nonzero, then that means there's conflicts
+	if err != nil {
+		if exitError, ok := err.(*exec.ExitError); ok {
+			exitStatus = exitError.ProcessState.ExitCode()
+		} else {
+			return -1, errors.New(tr.Tr.Get("failed to run merge program %q: %s", formattedMergeProgram, err))
+		}
+	}
+
+	outputFp, err := os.OpenFile(outputFile, os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		return -1, err
+	}
+
+	filename := replacements["D"]
+
+	stat, err := os.Stat(filename)
+	if err != nil {
+		return -1, err
+	}
+
+	inputFp, err := os.OpenFile(filename, os.O_RDONLY|os.O_CREATE, 0600)
+	if err != nil {
+		return -1, err
+	}
+
+	gf := lfs.NewGitFilter(cfg)
+	_, err = clean(gf, outputFp, inputFp, filename, stat.Size())
+	if err != nil {
+		return -1, err
+	}
+
+	return exitStatus, nil
+}
+
+func mergeCleanup(replacements map[string]string) {
+	ids := []string{"A", "O", "B", "D"}
+	for _, id := range ids {
+		os.Remove(replacements[id])
+	}
+}
+
+func mergeProcessInput(gf *lfs.GitFilter, filename string, replacements map[string]string, tag string) {
+	file, err := lfs.TempFile(cfg, fmt.Sprintf("merge-driver-%s", tag))
+	if err != nil {
+		Exit(tr.Tr.Get("could not create temporary file when merging: %s", err))
+	}
+	defer file.Close()
+	replacements[tag] = file.Name()
+	if len(filename) != 0 {
+		pointer, err := lfs.DecodePointerFromFile(filename)
+		if err != nil {
+			if errors.IsNotAPointerError(err) {
+				if err := lfs.CopyFileContents(cfg, filename, file.Name()); err != nil {
+					os.Remove(file.Name())
+					Exit(tr.Tr.Get("could not copy non-LFS content when merging: %s", err))
+				}
+			} else {
+				os.Remove(file.Name())
+				Exit(tr.Tr.Get("could not decode pointer when merging: %s", err))
+			}
+		}
+		cb, fp, err := gf.CopyCallbackFile("download", file.Name(), 1, 1)
+		if err != nil {
+			os.Remove(file.Name())
+			Exit(tr.Tr.Get("could not create callback: %s", err))
+		}
+		_, err = gf.Smudge(file, pointer, file.Name(), true, getTransferManifestOperationRemote("download", cfg.Remote()), cb)
+		fp.Close()
+	}
+}
+
+func init() {
+	RegisterCommand("merge-driver", mergeDriverCommand, func(cmd *cobra.Command) {
+		cmd.Flags().StringVarP(&mergeDriverAncestor, "ancestor", "", "", "file with the ancestor version")
+		cmd.Flags().StringVarP(&mergeDriverCurrent, "current", "", "", "file with the current version")
+		cmd.Flags().StringVarP(&mergeDriverOther, "other", "", "", "file with the other version")
+		cmd.Flags().StringVarP(&mergeDriverOutput, "output", "", "", "file with the output version")
```

Comment:
IIUC, this `git-lfs-merge` program will be the merge driver from Git's perspective, and then we'll invoke either the default `git-merge-file` or a user-specified program.  Either way, as the "actual" or outermost merge driver, according to the [docs](https://www.git-scm.com/docs/gitattributes#_defining_a_custom_merge_driver) it seems like it's our responsibility to leave the output in the same location as the current version input file:
> The merge driver is expected to leave the result of the merge in the file named with `%A` by overwriting it, and exit with zero status if it managed to merge them cleanly, or non-zero if there were conflicts.

If that's the case, could we drop the `--output` option (or equivalent fourth argument) entirely?  In the example in the manual page below, it shows:
```
git config merge.lfs-text.driver 'git lfs merge-driver --ancestor %O --current %A --other %B --marker-size %L --output %A
```
If the user is effectively always required to specify `%A` for both `--current` and `--output`, perhaps we can simplify things by skipping the latter and just outputting the `clean()`ed result back to the `mergeDriverCurrent` filename?

Reply:
We can, but I really don't like special-casing `--current` like that.  In my view, it's a bug that Git has decided to specify the driver like that, since it makes it hard to reason about the behaviour.  It also makes it hard to integrate with other programs which may want to use this outside of Git, which I'd like to keep open as a possibility for scripting purposes.